lib/IDE/Refactoring.cpp

```diff
@@ -5373,28 +5583,54 @@ class AsyncConverter : private SourceEntityWalker {
     }
   }
 
-  // TODO: Check for clashes with existing names and add all decls, not just
-  //       params
+  /// Add a mapping from each passed parameter to a new name, possibly
+  /// synthesizing a new one if hoisting it would cause a redeclaration or
+  /// shadowing. If there's no bound name and \c AddIfMissing is false, no
+  /// name will be added.
   void prepareNames(const ClassifiedBlock &Block,
                     ArrayRef<const ParamDecl *> Params,
                     bool AddIfMissing = true) {
-    for (auto *Param : Params) {
-      StringRef Name = Block.boundName(Param);
-      if (!Name.empty()) {
-        Names[Param] = Name.str();
-        continue;
-      }
+    for (auto *PD : Params) {
+      StringRef Name = Block.boundName(PD);
+      if (Name.empty() && AddIfMissing)
+        Name = PD->getNameStr();
+      assignName(PD, Name);
+    }
 
-      if (!AddIfMissing)
-        continue;
+    for (auto &Entry : Block.aliases()) {
+      auto It = Names.find(Entry.second);
+      assert(It != Names.end() && "Param should already have an entry");
+      Names[Entry.first] = It->second;
+    }
+  }
 
-      auto ParamName = Param->getNameStr();
-      if (ParamName.startswith("$")) {
-        Names[Param] = "val" + ParamName.drop_front().str();
-      } else {
-        Names[Param] = ParamName.str();
-      }
+  StringRef assignName(const Decl *D, StringRef Name) {
+    auto *VD = dyn_cast<VarDecl>(D);
+    if (!VD)
+      return StringRef();
+
+    if (Name.empty() || Name == StringRef("_"))
+      return StringRef();
+
+    llvm::SmallString<32> NewName;
+    if (Name.startswith("$")) {
+      NewName.append("val");
+      NewName.append(Name.drop_front());
+    } else {
+      NewName.append(Name);
     }
+
+    size_t BeforeUniqueSize = NewName.size();
```

Comment:
Wouldn’t it be easier to just store `NewName` here and reset it in the loop if necessary?

Like
```cpp
auto BeforeAppendingUniqueSuffix = NewName;
while (…) {
  NewName = BeforeAppendingUniqueSuffix;
  NewName.append(…);
   ...
}
```

Would avoid the `pop_back_n` dance and just keeps track of less state, I think.

Reply:
I was just avoiding the copying here, but I suppose it's not really performance critical here.